eng/scripts/TypeSpec-Validation.ps1

```diff
@@ -4,51 +4,113 @@ param (
   [switch]$GitClean = $false,
   [switch]$DryRun = $false,
   [string]$BaseCommitish = "HEAD^",
-  [string]$TargetCommitish = "HEAD"
+  [string]$TargetCommitish = "HEAD",
+  [int]$FolderCount = 0,
+  [int] $Parallelism = 6
 )
 
 . $PSScriptRoot/Logging-Functions.ps1
 . $PSScriptRoot/Suppressions-Functions.ps1
 
-$typespecFolders, $checkedAll = &"$PSScriptRoot/Get-TypeSpec-Folders.ps1" -BaseCommitish:$BaseCommitish -TargetCommitish:$TargetCommitish -CheckAll:$CheckAll
+$typespecFolders, $checkedAll = &"$PSScriptRoot/Get-TypeSpec-Folders.ps1" `
+  -BaseCommitish:$BaseCommitish `
+  -TargetCommitish:$TargetCommitish `
+  -CheckAll:$CheckAll
+
+# Use -FolderCount to do performance testing on Typespec-Validation.ps1 with 
+# -CheckAll set to $true. This constrains the typeSpec folders to the first N 
+# folders. 
+if ($FolderCount) { 
+  $typespecFolders = $typespecFolders[0..($FolderCount - 1)]
+}
 
 $typespecFoldersWithFailures = @()
 if ($typespecFolders) {
-  $typespecFolders = $typespecFolders.Split('',[System.StringSplitOptions]::RemoveEmptyEntries)
-  foreach ($typespecFolder in $typespecFolders) {
-    LogGroupStart "Validating $typespecFolder"
+  # Group typespec folders by "service" (the first folder in the path). This
+  # parallel validation at the level of the service.
+  # Example: 'specification/ai/DocumentIntelligence' -> 'specification/ai'.
+  $serviceFolders = $typespecFolders | Group-Object -Property { 
+    # "specification\ai\DocumentIntelligence" -> @("specification", "ai", "DocumentIntelligence")
+    $splitPath = $_ -replace '\\', '/' -split '/'
+
+    # @("specification", "ai", "DocumentIntelligence") -> "specification/ai"
+    return $splitPath[0..1] -join '/'
+  }
+  
+  Write-Host "Starting per-service parallel validation with parallelism: $Parallelism"
 
-    if ($checkedAll) {
-      $suppression = Get-Suppression "TypeSpecValidationAll" $typespecFolder
-      if ($suppression) {
-        $reason = $suppression["reason"] ?? "<no reason specified>"
-        LogInfo "Suppressed: $reason"
-        LogGroupEnd
+  $serviceFolders `
+  | ForEach-Object -ThrottleLimit $Parallelism -Parallel {
```

Comment:
Does it make sense to parallelize across agents to avoid race conditions and things stomping on each other? Similar to what we do for generating a job matrix?

Reply:
We can do this as a next step, but we'd like to get as much perf as we can from parallelism on a single machine, since it will also help scenarios like validating all specs locally.